src/ngAnimate/animation.js

```diff
@@ -187,7 +187,7 @@ var $$AnimationProvider = ['$animateProvider', /** @this */ function($animatePro
           var element = animationEntry.from ? animationEntry.from.element : animationEntry.element;
           var extraClasses = options.addClass;
           extraClasses = (extraClasses ? (extraClasses + ' ') : '') + NG_ANIMATE_CLASSNAME;
-          var cacheKey = $$animateCache.cacheKey(node, event, extraClasses, options.removeClass);
+          var cacheKey = $$animateCache.cacheKey(node, animationEntry.event, extraClasses, options.removeClass);
```

Comment:
I'm still unable to isolate the steps which lead to the issue. While trying to find out I managed to get why it leaves on the containsCachedAnimationWithoutDuration() check. I can see that there is a leave and an enter animation getting triggered. Both are in the animation queue, but when code goes through the animations it creates the cache key based on the first event (leave in this case). When it later comes to the actual execution of the animation it first handles the leave, containsCachedAnimationWithoutDuration() returns false as this key is not in the cache yet and then adds it to cache later (with !valid as the transition-duration is null). Then later the same happens for the enter animation, but due to the same cache key as the leave animation containsCachedAnimationWithoutDuration() now finds the existing non valid entry and returns true and the enter is skipped.

Any idea how the duration can be null?

Reply:
So I managed now to reproduce it

http://plnkr.co/edit/6CsoTul8LBIGBcHfioYJ
(hitting action, when text goes red it means we are into the situation).

So you need a directive and changing ng-switch and changing ng-class as far as I see. No idea how to put that into a test.

Reply:
Thanks, I'll take a look. I take it the custom directive is necessary? replace: true always complicates things.

Reply:
Any chance that one can trigger the animation events for an element in the tests? Might be simpler than rebuilding the complex structure.

Reply:
There's animationSpec.js, which tests this behavior somewhat in isolation: https://github.com/angular/angular.js/blob/a5537359c531ca9c32ea481f9189e9d4354859c0/test/ngAnimate/animationSpec.js#L551-L571
But since the cacheKey is created in animateCss, I'm not sure it's possible to recreate the case here.
Maybe we'll have to build a full integration test including the custom directive first.

Reply:
I managed to trim down your plnkr to what I think is the minimum to reproduce the problem: http://plnkr.co/edit/m5CaegjmdtpR1pUX7zZ8?p=preview I'll try to put this into a test, then it should be easier to see what happens when. Interestingly, the ngClass directive is necessary for the bug to appear.

> Any idea how the duration can be null?

ngAnimate is reading possible animation styles from the element during animation. So if you have an element / directive that could technically be animated, it will look on the classes to find animations. If none are found (basically the default), the animation duration is null. Or did you mean null as in null and not 0?

Reply:
I've opened a PR #16681 that simplifies the fix and adds tests